spring-integration-core/src/main/java/org/springframework/integration/handler/advice/RequestHandlerRateLimiterAdvice.java

```
@@ -0,0 +1,114 @@
+import org.springframework.messaging.MessagingException;
+import org.springframework.util.Assert;
+
+/**
+ * Rate limiter advice, which will ensure call to service is Rate limited within set rate limit.
+ * Request calls breaching the rate limit will wait for time rate limit is not breached.
+ *
+ * @author Meherzad Lahewala
+ * @since 5.1
+ */
+public class RequestHandlerRateLimiterAdvice extends AbstractRequestHandlerAdvice {
+
+	private final AtomicInteger currentIndex;
+	private final long maxDelayInMillis;
+	private final long oneTimeUnitInMillis;
+	private final AtomicLongArray queue;
+	private final int rate;
+
+	/**
+	 * Create an instance of RequestHandlerRateLimiterAdvice with the provided arguments.
+	 *
+	 * @param rate the rate per time unit.
+	 * @param maxDelay the max allowable delay in time units. Use -1 for no max delay.
+	 * @param timeUnit the time unit.
+	 */
+	public RequestHandlerRateLimiterAdvice(int rate, int maxDelay, TimeUnit timeUnit) {
+		Assert.isTrue(rate > 0, "Rate should be a positive integer");
+		Assert.notNull(timeUnit, "Timeunit should be non null");
+		this.currentIndex = new AtomicInteger(0);
+		this.maxDelayInMillis = timeUnit.toMillis(maxDelay);
+		this.oneTimeUnitInMillis = timeUnit.toMillis(1);
+		this.queue = new AtomicLongArray(rate);
+		this.rate = rate;
+	}
+
+	/**
+	 * Create an instance of RequestHandlerRateLimiterAdvice with the provided arguments.
+	 *
+	 * @param rate the rate per time unit.
+	 * @param timeUnit the time unit.
+	 */
+	public RequestHandlerRateLimiterAdvice(int rate, TimeUnit timeUnit) {
+		this(rate, -1, timeUnit);
+	}
+
+	@Override
+	protected Object doInvoke(ExecutionCallback callback, Object target, Message<?> message) throws Exception {
+		delayIfNecessary(target, message);
+		return callback.execute();
+	}
+
+	private void delayIfNecessary(Object target, Message<?> message) throws InterruptedException {
+		long timeToWait = -1;
+		while (timeToWait == -1) {
```

I believe there should be a way for fail fast for a single call or as we already know per `rate`. 
I mean if we still in the `rate` we doin't need to sleep or even calculate `System.currentTimeMillis()` at all.

I need to understand algorithm better, because it looks like I have more questions, than review requires 😢